File: src/PSD.py
import numpy as np
import scipy.signal as sp
from scipy.fftpack import fft

def find_nearest(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return idx

def psd(data, window_factor = 10, overlap_factor = 20):
    """
    Returns the frequency and intensity of the power spectral density of `data`.

    inputs :
    ----------------------
    data: The velocity timeseries
    window_factor : this parameter decides the length of the window by dividing the total length of the data by 'window_factor"
                    defaults to 10.
    overlap_factor : decides the overlap between windows. Works like window_factor by dividing the total length. defaults to 20 meaning, that the PSD is averaged 19 times.
    (upper=1, lower=1/8) : defines the range to fit the powerlaw in the PSD (as multiples of the quarter note).

    returns :
    ----------------------
    (f, pxx) : A 2D numpy aray containing the frequencies and the associated intensity of the psd
    (l,u,fit) : A 2D array of respectively the lower limit of the ci, THE UPPER LIMIT OF THE ci and the fit itself for the power-law.
    the minimal duration, which can and should be used to compute the period equivalent to the frequencies. The values are typically 240 (16th notes) or 320 (triplet).
    """
    wdw, nov = int(len(data)/window_factor), int(len(data)/overlap_factor)
    f, pxx = sp.welch(data, nperseg = wdw, noverlap = nov, detrend = "constant", window = "hanning")
    return f, pxx

def fit_powerlaw(f, pxx, mindur, tpq, upper = 1, lower = 0.125):
    """
    inputs :
    ----------------------
    (f, pxx) : the result of the `psd` function.
    (mindur, tpq) : minimum duration (in ticks) of the grid points, and ticks per quarter note.
    (upper=1, lower=1/8) : defines the range to fit the powerlaw in the PSD (as multiples of the beat).

    returns:
    (f, fit) the fit and its frequencies (ready to plot)
    """
    # Convert beats to frequencies
    flower = mindur / upper / tpq
    fupper = mindur / lower / tpq
    start, stop = find_nearest(f, flower), find_nearest(f, fupper)

    f = f[start:stop]
    fit = np.polyfit(np.log10(f),np.log10(pxx[start:stop]),1)
    fit_data = 10**(np.log10(f)*fit[0]+fit[1])

    slope = (np.log10(fit_data[-1]) - np.log10(fit_data[0]))/(np.log10(f[-1]) - np.log10(f[0]))

    return f, fit_data, slope

def chisquare_ci(N = 19):
    """
    Retrieves the critical z-value for the chi-square distribution.
    This is used during PSD estimation to get a 95% confidence interval (CI).

    input:
    ------------------------
    N : the number of averaging used to produce the estimation.
    (argument `overlap` of `psd` minus 1)

    returns:
    ------------------------
    lower, upper, N

    The PSD should be transformed as p_low = p*a1/N and p_up = p*a2/N to give the
    range of the 95% CI values.
    """
    lower = []
    upper = []
    if N == 19:
        a1 = 8.907
        a2 = 32.852
    elif N == 17:
        a1 = 7.56
        a2 = 30.19
    elif N == 4:
        a1 = 0.485
        a2 = 11.14
    elif N == 3:
        a1 = 0.216
        a2 = 9.348
    elif N == 2:
        a1 = 0.0506
        a2 = 7.378
    elif N == 1:
        a1 = 0.000982
        a2 = 5.024
        # for p in pxx:
        #     lower.append(p*a1/N)
        #     upper.append(p*a2/N)
    else:
        raise Exception("Unknown overlap")
    return a1, a2, N

File: src/PSD.jl
using DrWatson
@quickactivate "VelocityMTD"

using PyCall

pushfirst!(PyVector(pyimport("sys")."path"), srcdir())
pypsd = pyimport("PSD")
pyimport("importlib")."reload"(pypsd)

"""
Plot the PSD and its fit on `ax`. Input is velocity timeseries (or the notes
to calculate the timeseries from, by averaging velocity).

`kwargs` are propagated into `Python.psd`, while `minsub, tpq` are
"""
function psdfit!(ax, notes::Notes, middle = nothing;
    minsub = 12, tpq = 960, kwargs...)

    t, v = timeseries(notes, :velocity, mean, 0:(1//minsub):1)
    v = [ismissing(vi) ? 0 : vi for vi in v]
    psdfit!(ax, v, middle; minsub = minsub, tpq = tpq, kwargs...)
end

function psdfit!(ax, v::Vector, middle = nothing;
    minsub = 12, tpq = 960, kwargs...)

    mindur = tpq÷minsub

    f2b = (f) -> @. mindur/f/tpq
    b2f = (b) -> @. mindur/b/tpq

    f, psd = pypsd.psd(v; kwargs...)
    lo, up, N = pypsd.chisquare_ci(get(kwargs, :overlap_factor, 20) - 1)

    beats = f2b.(f[2:end])

    ax.loglog(beats, psd[2:end], c= "C0", lw = 1.5, basex=2)
    ax.fill_between(beats, psd[2:end] .* up ./ N, psd[2:end] .* lo ./ N, color = "C0", alpha = 0.25)

    if isnothing(middle)
        fs1, sl1, slope1 = pypsd.fit_powerlaw(f, psd, mindur, tpq; upper = maximum(beats), lower = minimum(beats))
        ax.plot(f2b.(fs1), sl1, label = "\$\\beta\$ = $(round(slope1; digits=2))", c = "C5")
        slope2 = 0.0
    else
        fs1, sl1, slope1 = pypsd.fit_powerlaw(f, psd, mindur, tpq; upper = middle, lower = minimum(beats))
        fs2, sl2, slope2 = pypsd.fit_powerlaw(f, psd, mindur, tpq; upper = maximum(beats), lower = middle)
        ax.plot(f2b.(fs1), sl1, label = "\$\\delta\$ = $(round(slope1; digits=2))", c = "C2")
        ax.plot(f2b.(fs2), sl2, label = "\$\\beta\$ = $(round(slope2; digits=2))", c = "C5")
    end

    ax.legend()
    ax.set_xlabel("period (beats)")
    ax.grid()
    return slope1, slope2
end
